Introduce RecipientData and use it to allow abandon failed payments

shaavan opened the pull request:
<p><strong>Resolves #2837</strong></p>

<h3>Summary</h3>

<p>This PR introduces a new field, <code>RecipientData</code>, in the <code>ReceiveTlvs</code> struct. The key changes are as follows:</p>

<ol>
  <li>
    <strong>Introduction of <code>RecipientData</code>:</strong>
    <ul>
      <li>This struct allows users to append additional data to the created <code>reply_path</code>, which they expect to receive back from the counterparty along with their response.</li>
    </ul>
  </li>
  <li>
    <strong>Optional <code>PaymentId</code> in <code>RecipientData</code>:</strong>
    <ul>
      <li><code>RecipientData</code> includes an optional <code>PaymentId</code> field that can be utilized by <code>OffersMessage</code>.</li>
      <li>This feature enables users to abandon outbound payments that fail for any reason.</li>
    </ul>
  </li>
  <li>
    <strong>Integration with <code>create_blinded_path</code> flow:</strong>
    <ul>
      <li>The PR integrates <code>RecipientData</code> into the <code>create_blinded_path</code> process.</li>
      <li>This enhancement is utilized in the <code>pay_for_offer</code> and <code>create_refund_builder</code> functions to pass <code>PaymentId</code> along with the generated <code>reply_path</code> and path respectively.</li>
    </ul>
  </li>
</ol>

<p>These improvements aim to allow passing additional information with the create <code>reply_path</code> and managing outbound payment failures more robustly.</p>

**Note:**
This PR builds on #2996

shaavan commented:
Updated from [pr3085.01](https://github.com/shaavan/rust-lightning/commits/pr3085.01) to [pr3085.02](https://github.com/shaavan/rust-lightning/commits/pr3085.02) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3085.01..pr3085.02)):
Addressed @tnull comment

Changes:
1. Rebase on main to resolve merge conflicts.

shaavan commented:
Updated from [pr3085.02](https://github.com/shaavan/rust-lightning/commits/pr3085.02) to [pr3085.03](https://github.com/shaavan/rust-lightning/commits/pr3085.03) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3085.02..pr3085.03)):

Changes:
1. Fix Fuzz failure

shaavan commented:
There’s another variation of this PR that might help us achieve our goal using enums. You can check it out here: [link to branch](https://github.com/shaavan/rust-lightning/commits/issue2837-4) (check out the top three commits).

This approach will help keep the `RecipientData` for `Offers Messages` and `Custom Messages` separate, making the use of `RecipientData` more focused and intentional.

Let me know which of the two approaches seems better.

jkczyz commented:
> There’s another variation of this PR that might help us achieve our goal using enums. You can check it out here: [link to branch](https://github.com/shaavan/rust-lightning/commits/issue2837-4) (check out the top three commits).
> 
> This approach will help keep the `RecipientData` for `Offers Messages` and `Custom Messages` separate, making the use of `RecipientData` more focused and intentional.
> 
> Let me know which of the two approaches seems better.

@valentinewallace @TheBlueMatt Would you mind checking out the two alternatives for including `RecipientData` in blinded paths? This PR uses a single struct which is passed to each message handler. The alternative approach in the linked branch uses an enum such that there is a different variant for each handler. Each variant wraps another type, which is passed to the handler instead of `RecipientData`.

jkczyz commented:
> Hmm, yea, we need to figure out what we want to do with `ReceiveTlvs` as a general matter. We currently don't use it much, and hid the `path_id` from LDK devs because we weren't sure if/how we wanted to expose it to users to set their own. I guess if we took `ReceiveTlvs` and defined some "LDK internal range" of types we could use an enum (or private struct) for those and then have a `custom_other_tlvs` field that we make pub?

Presumably, the other TLVs would be `Vec<(u64, Vec<u8>)>`? Could be another variant of that enum. Not sure if a private struct would be much better since we eventually need to pass the data to the handler.

> Do we currently use the `path_id` at all internally to LDK?

No, I believe we don't set it at all. Last we spoke about it, the thought was to reserve it for internal use.

jkczyz commented:
> I wonder if we should bother? Like, there's nothing special about the TLV the spec says should be the path_id, we can ignore it and no one will ever know. That would simplify the types here a bit.

Probably fine to ignore it or just encode the `message::ReceiveTlvs` enum (assuming this replaces `RecipientData`) as the `path_id`. Then we'd essentially have our own "namespace" for internal/custom types. Though this would diverge from what we do for `payment::ReceiveTlvs`.

jkczyz commented:
Checking my understanding regarding `path_id`. The spec (BOLT4) doesn't distinguish whether it's used for messages or payments, but it seems to imply either.

When a recipient creates a blinded path to itself:

> - MAY store private data in `encrypted_data_tlv[r].path_id` to verify that the route is used in the right context and was created by them

The final recipient (payment or message not specified, but message is then implied):

> - MUST ignore the message if the `path_id` does not match the blinded route it created

But then in the rational, the spec uses a payment example:

> The final recipient must verify that the blinded route is used in the right
> context (e.g. for a specific payment) and was created by them. Otherwise a
> malicious sender could create different blinded routes to all the nodes that
> they suspect could be the real recipient and try them until one accepts the
> message. The recipient can protect against that by storing `E_r` and the
> context (e.g. a `payment_hash`), and verifying that they match when receiving
> the onion. Otherwise, to avoid additional storage cost, it can put some private
> context information in the `path_id` field (e.g. the `payment_preimage`) and
> verify that when receiving the onion. Note that it's important to use private
> information in that case, that senders cannot have access to.

Later on in the BOLT under the "Onion Messages" section, the writer of `onionmsg_tlv`:

> - For the final node's `onionmsg_tlv`:
>   - if the final node is permitted to reply:
>     - For every `reply_path` `path`:
>       - MAY use `path_id` to contain a secret so it can recognize use of this `reply_path`.

And then the reader:

> - otherwise (it is the final node):
>   - if `path_id` is set and corresponds to a path the reader has previously published in a `reply_path`:
>     - if the onion message is not a reply to that previous onion:
>       - MUST ignore the onion message
>   - otherwise (unknown or unset `path_id`):
>     - if the onion message is a reply to an onion message which contained a `path_id`:
>       - MUST respond (or not respond) exactly as if it did not send the initial onion message.

Then in the rational:

> Care must be taken that replies are only accepted using the exact
> reply_path given, otherwise probing is possible.  That means checking
> both ways: non-replies don't use the reply path, and replies always
> use the reply path.

Note how that section talks about reply paths.

In LDK, we don't use `path_id` currently.
- `payment::ReceiveTlvs` doesn't have a `path_id`, but instead has `payment_secret` and `payment_context` fields encoded in custom TLVs.
- `message::ReceiveTlvs` does have a `path_id` field, but it is an `Option` and is never set.

IIUC, we shouldn't include a `path_id` in `Offer::paths` or `Refund::paths` as it would increase the QR code size. But for reply paths and (possibly) payment paths, we should. If we encoded the `RecipientData` in the `path_id` (or any TLV really), we could use its presence and variant to determine if it is being used by the appropriate handler.

I think that (plus any additional handler checks) would be sufficient to prevent de-anonymization discussed in the quoted rationale sections. @TheBlueMatt @valentinewallace Does that seem to check out?

If so, we could drop the `path_id` field as discussed in my previous comment and still use the `path_id` TLV for encoding `RecipientData`. I'm not sure if we want to replace `RecipientData` with `ReceiveTlvs` though. This would prevent us from adding custom TLVs in the future. Or at very least would require refactoring the code back to how it currently is in this PR. Thoughts?

jkczyz commented:
> Is the attack there basically that someone can take an offer we generated, then request the invoice_request from every node in the network and see who responds (or equivalent for payments)?

Really to any node within N hops of the introduction node in one of the offer paths. Though using the `path_id` in an offer path would increase the QR code size. The attack seems only applicable when the recipient is an announced node. Otherwise, the attacker can't create a blinded path without a suspected node id to try for the recipient.

jkczyz commented:
> Hmm, okay, I think we're talking about different things, then. I think my above-described attack works though, right? Currently anyone can probe all public nodes and figure out if it is the node that issued an LDK offer, I think?

Yeah, I was just saying you could limit the search.

> Right, though it seems like if you know the node's pubkey you could do it too? Seems like something we should fix in both cases.

Correct. IIUC, the fix would result in an increased offer QR code size.

jkczyz commented:
> In any case, as it relates to this PR, ISTM we don't care too much about users using `path_id` in their own blinded paths for their own use, so we might as well just have a top-level enum which users can put whatever they want in and then have an LDK internal variant that has private fields that only LDK can use.

Not sure I understand the suggestion here. We want different variants for *each* LDK usage so we can tell if the path is used correctly. Having a single variant with a bunch of `Option` fields makes this less intuitive. Or are you saying it would be a two-level enum? Feels like we should just have one enum with all the LDK internal variants plus one variant for user-specific usage (i.e., to use with `CustomOnionMessageHandler`) and never actually expose the enum publicly.

jkczyz commented:
> Right, we're on the same page, my only point is that it seems like we should have a single-level enum, whereas the previous suggestion was a two-level enum.

I think the previous suggestion was a struct (`ReceiveTlvs`) holding an enum (`RecipientData`). 😄  IIUC, that is what we would need to hide the LDK-specific variants from the user. Only instead of the `BlindedPath` creation methods taking `RecipientData`, they would take `ReceiveTlvs` (just like the payment variations only with non-pub field). `ReceiveTlvs` would need a constructor to create one holding a custom variant, while LDK code could freely construct `ReceiveTlvs` holding the other `RecipientData` variants as needed.

jkczyz commented:
> Or we could go the other way - have a top-level enum where the LDK-hidden variants contain a struct, that way the user-visible API is simpler.

SGTM. Each variant will likely need to hold a struct in order have something to pass to each handler. But, yeah, those structs can be `pub` only not creatable externally, IIUC. So the user would only be able to create the variant for custom data.

shaavan commented:
Updated from [pr3085.03](https://github.com/shaavan/rust-lightning/commits/pr3085.03) to [pr3085.04](https://github.com/shaavan/rust-lightning/commits/pr3085.04) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3085.03..pr3085.04)):

Addressed comments from @TheBlueMatt and @jkczyz

Changes:

1. Rebase on main.
1. Updated `ReceiveTlvs` to include the optional `RecipientData` enum.
2. This new approach allows for the serialization of the entire `ReceiveTlvs` under type `6`.
3. Using an enum allows us to ensure the response message received over the appended `reply_path` is of the correct type before invoking the handler.

shaavan commented:
Updated from [pr3085.04](https://github.com/shaavan/rust-lightning/commits/pr3085.04) to [pr3085.05](https://github.com/shaavan/rust-lightning/commits/pr3085.05) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3085.04..pr3085.05)):

Changes:
1. Corrected Readable/Writable implementation for ControlTlvs.

jkczyz commented on lightning/src/blinded_path/message.rs:
```diff
@@ -74,16 +68,106 @@ impl Writeable for ForwardTlvs {
 	}
 }
 
+/// Similar to [`ForwardTlvs`], but these TLVs are for the final node.
+pub(crate) struct ReceiveTlvs {
+	pub tlvs: Option<RecipientData>
+}
```
Let's move this back where it was to keep the blame log clean.

shaavan replied:
Done! [8e4e1aa](https://github.com/lightningdevkit/rust-lightning/commit/8e4e1aabadb5fda9f4d0ec2c845263c94c273643)
Thanks!

jkczyz commented on lightning/src/blinded_path/message.rs:
```diff
@@ -74,16 +68,106 @@ impl Writeable for ForwardTlvs {
+	pub tlvs: Option<RecipientData>
+}
+
+impl ReceiveTlvs {
+	pub fn new(tlvs: Option<RecipientData>) -> Self {
+		ReceiveTlvs { tlvs }
```
No need for a constructor, IMO.

shaavan replied:
I can agree!
This also help keep the code symmetrical with `ForwardTlvs`
Done! [8e4e1aa](https://github.com/lightningdevkit/rust-lightning/commit/8e4e1aabadb5fda9f4d0ec2c845263c94c273643)
Thanks!

jkczyz commented on lightning/src/blinded_path/message.rs:
```diff
@@ -74,16 +68,106 @@ impl Writeable for ForwardTlvs {
+	}
+}
+
+/// Represents additional data appended along with the sent reply path.
```
This is not restricted to reply paths.

s/appended .../included by the recipient in a [`BlindedPath`].

shaavan replied:
Done! [8e4e1aa](https://github.com/lightningdevkit/rust-lightning/commit/8e4e1aabadb5fda9f4d0ec2c845263c94c273643)
Thanks!

jkczyz commented on lightning/src/blinded_path/message.rs:
```diff
@@ -74,16 +68,106 @@ impl Writeable for ForwardTlvs {
+
+/// Represents additional data appended along with the sent reply path.
+///
+/// This data can be utilized by the final recipient for further processing
+/// upon receiving it back.
```
Let's rephrase this slightly:

This data is encrypted and will be included when sending through a [`BlindedPath`], allowing the recipient to authenticate the message.

shaavan replied:
Built upon your suggestion! [8e4e1aa](https://github.com/lightningdevkit/rust-lightning/commit/8e4e1aabadb5fda9f4d0ec2c845263c94c273643)
Also added a brief comment that the data can be used for further processing.

jkczyz commented on lightning/src/blinded_path/message.rs:
```diff
@@ -74,16 +68,106 @@ impl Writeable for ForwardTlvs {
+}
+
+impl RecipientData {
+	/// Creates a new RecipientData::OffersContext instance.
+	pub fn new_for_offers(payment_id: Option<PaymentId>) -> Self {
+		RecipientData::OffersContext(OffersData {
+			payment_id,
+		})
+	}
```
Let's drop this constructor since (1) we don't want users to be able to create this (2) we'll likely have other variations in the future, and (3) the struct's fields are `pub` so it can be constructed without one.

jkczyz replied:
Actually, we'll have to make the fields `pub(crate)` to make ensure (1).

shaavan replied:
Removed the redundant constructor in [8e4e1aa](https://github.com/lightningdevkit/rust-lightning/commit/8e4e1aabadb5fda9f4d0ec2c845263c94c273643)

Regarding the visibility of the data, I think we'll need to keep it as `pub`. Since `RecipientData` is a field within `PeeledOnion`, which itself is a public struct, it seems like restricting the internal fields' visibility any further might not be possible.

```rust
pub enum PeeledOnion<T: OnionMessageContents> {
	///
	Receive(ParsedOnionMessageContents<T>, Option<RecipientData>, Option<BlindedPath>)
}```

Let me know what you think about it!

jkczyz replied:
Seems like making `payment_id` have `pub(crate)` visibility works. Hmmm... but I guess that would restrict who could use that to implement `OffersMessageHandler`. So maybe we'll need to keep it as you have it now afterall.

jkczyz commented on lightning/src/blinded_path/message.rs:
```diff
@@ -74,16 +68,106 @@ impl Writeable for ForwardTlvs {
+		})
+	}
+
+	/// Creates a new RecipientData::CustomContext instance.
+	pub fn new_for_custom(data: Vec<u8>) -> Self {
+		RecipientData::CustomContext(data)
+	}
```
We can probably drop this as well.

shaavan replied:
Done! [8e4e1aa](https://github.com/lightningdevkit/rust-lightning/commit/8e4e1aabadb5fda9f4d0ec2c845263c94c273643)
Thanks!

jkczyz commented on lightning/src/blinded_path/message.rs:
```diff
@@ -74,16 +68,106 @@ impl Writeable for ForwardTlvs {
+	}
+}
+
+impl Writeable for RecipientData {
+	fn write<W: Writer>(&self, writer: &mut W) -> Result<(), io::Error> {
+		match self {
+			RecipientData::OffersContext(offers_data) => {
+				1u8.write(writer)?; // Identifier for OffersContext
+				offers_data.payment_id.write(writer)?; // Write the payment_id
+			}
+			RecipientData::CustomContext(data) => {
+				2u8.write(writer)?; // Identifier for CustomContext
+				data.write(writer)?; // Write the custom data
+			}
+		}
+		Ok(())
+	}
+}
+
+impl Readable for RecipientData {
+	fn read<R: io::Read>(r: &mut R) -> Result<Self, DecodeError> {
+		let type_id = u8::read(r)?;
+		match type_id {
+			1u8 => {
+				let payment_id = Option::<PaymentId>::read(r)?;
+				Ok(RecipientData::new_for_offers(payment_id))
+			}
+			2u8 => {
+				let data = Vec::<u8>::read(r)?;
+				Ok(RecipientData::new_for_custom(data))
+			}
+			_ => Err(DecodeError::InvalidValue),
+		}
+	}
+}
```
Could you use `impl_writeable_tlv_based_enum` on `RecipientData` and `impl_writeable_tlv_based` on `OffersData`?

shaavan replied:
Used the macros to simply the code in [8e4e1aa](https://github.com/lightningdevkit/rust-lightning/commit/8e4e1aabadb5fda9f4d0ec2c845263c94c273643)
Thanks a lot for the pointers!

jkczyz commented on lightning/src/blinded_path/message.rs:
```diff
@@ -74,16 +68,106 @@ impl Writeable for ForwardTlvs {
 
+/// Similar to [`ForwardTlvs`], but these TLVs are for the final node.
+pub(crate) struct ReceiveTlvs {
+	pub tlvs: Option<RecipientData>
```
Let's keep this as `path_id` to align with the spec.

shaavan replied:
Done! [8e4e1aa](https://github.com/lightningdevkit/rust-lightning/commit/8e4e1aabadb5fda9f4d0ec2c845263c94c273643)

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -1414,7 +1425,8 @@ where
 		match self.peel_onion_message(msg) {
-			Ok(PeeledOnion::Receive(message, path_id, reply_path)) => {
+			Ok(PeeledOnion::Receive(message, _tlvs, reply_path)) => {
+				let path_id: Option<[u8; 32]> = None;
```
We'll want to drop the `path_id` entirely from `Responder` in a separate commit since it is not useful at this level. For background, see https://github.com/lightningdevkit/rust-lightning/pull/3078#discussion_r1636208042.

jkczyz commented on lightning/src/onion_message/packet.rs:
```diff
@@ -314,7 +314,7 @@ impl Readable for ControlTlvs {
 			(None, None) => None,
 		};
 
-		let valid_fwd_fmt = next_hop.is_some() && path_id.is_none();
```
Should this check stay?

shaavan replied:
I think we should keep this check. It's better to have a (potentially redundant?) check than to miss something important, right? :)
I've reverted the check to its original version in [8e4e1aa](https://github.com/lightningdevkit/rust-lightning/commit/8e4e1aabadb5fda9f4d0ec2c845263c94c273643).

Let me know what you think!

shaavan replied:
Understood! Kept the check intact in the current version of PR.

jkczyz commented on lightning/src/util/ser.rs:
```diff
@@ -103,6 +103,7 @@ impl Writer for LengthCalculatingWriter {
 	}
 }
 
+
```
Remove errant change.

shaavan replied:
Done! [8e4e1aa](https://github.com/lightningdevkit/rust-lightning/commit/8e4e1aabadb5fda9f4d0ec2c845263c94c273643)

jkczyz commented on lightning/src/util/ser_macros.rs:
```diff
@@ -557,6 +557,7 @@ macro_rules! _decode_tlv_stream_range {
 			// Finally, read the length and value itself:
 			let length: ser::BigSize = $crate::util::ser::Readable::read(&mut stream_ref)?;
 			let mut s = ser::FixedLengthReader::new(&mut stream_ref, length.0);
+
```
Likewise.

shaavan replied:
Done! [8e4e1aa](https://github.com/lightningdevkit/rust-lightning/commit/8e4e1aabadb5fda9f4d0ec2c845263c94c273643)

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -432,7 +433,7 @@ pub trait MessageRouter {
 		T: secp256k1::Signing + secp256k1::Verification
 	>(
-		&self, recipient: PublicKey, peers: Vec<PublicKey>, secp_ctx: &Secp256k1<T>,
+		&self, recipient: PublicKey, recipient_tlvs: Option<RecipientData>, peers: Vec<PublicKey>, secp_ctx: &Secp256k1<T>,
```
Here and elsewhere: s/recipient_tlvs/recipient_data

shaavan replied:
Done! [8e4e1aa](https://github.com/lightningdevkit/rust-lightning/commit/8e4e1aabadb5fda9f4d0ec2c845263c94c273643)
Thanks!

jkczyz commented on lightning/src/routing/router.rs:
```diff
@@ -178,17 +178,17 @@ impl< G: Deref<Target = NetworkGraph<L>> + Clone, L: Deref, ES: Deref, S: Deref,
 		T: secp256k1::Signing + secp256k1::Verification
 	> (
-		&self, recipient: PublicKey, peers: Vec<PublicKey>, secp_ctx: &Secp256k1<T>,
+		&self, recipient: PublicKey, recipient_tlvs: Option<RecipientData>, peers: Vec<PublicKey>, secp_ctx: &Secp256k1<T>,
```
Hmm... not sure if we need these in payment paths. @valentinewallace Don't we use some other mechanism for including custom data there?

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -10326,18 +10335,24 @@ where
 				}
 			},
 			OffersMessage::InvoiceError(invoice_error) => {
+				abandon_if_payment(offers_data.payment_id);
```
Could you update/add a test to `offers_tests.rs` that show that the payment is abandoned when this case is hit?

jkczyz replied:
Could you also use `expect_recent_payment` to show it is in `RecentPaymentDetails::Abandoned`? We should check these for all tests that result in this state.

shaavan replied:
Thanks for your suggestion!
Using `expect_recent_payment` sounds like a good idea, but it might not be very straightforward in this case!

Once `abandon_payment` is called, it removes the payment from `pending_outbound_payments` after generating the appropriate event. Here's the relevant [code snippet](https://github.com/lightningdevkit/rust-lightning/blob/87fc324c71500f017adf535316465825efe083f1/lightning/src/ln/outbound_payment.rs#L1771-L1776).

Given this, checking for the `InvoiceRequestFailed` event might be more fitting to confirm the payment failure.

Additionally, since `abandon_payment` generates an event, any test triggering this should have an unhandled pending event, leading to a failure. This actually happened with `fails_sending_invoice_without_blinded_payment_paths_for_offer` before adding the event handling.

The good news is that since no other tests are failing, it looks like no other tests are reaching this state.

Let me know what you think!

jkczyz replied:
Ah, gotcha.

shaavan commented:
Updated from [pr3085.05](https://github.com/shaavan/rust-lightning/commits/pr3085.05) to [pr3085.06](https://github.com/shaavan/rust-lightning/commits/pr3085.06) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3085.05..pr3085.06)):
Addressed @jkczyz comments

Changes:

1. Rearranged and refactored code and removed redundant constructors to keep code DRY and git blame clean.
2. Update the docs and variable name to be more precise and align with Spec.
3. Remove path_id from Responder in a separate commit.
4. `recipient_tlvs` -> `recipient_data`.

jkczyz commented on lightning/src/ln/offers_tests.rs:
```diff
@@ -1332,6 +1332,16 @@ fn fails_sending_invoice_without_blinded_payment_paths_for_offer() {
 	assert_eq!(invoice_error, InvoiceError::from(Bolt12SemanticError::MissingPaths));
+
+	// Confirm that david drops this failed payment from his pending outbound payments.
+	let events = david.node.get_and_clear_pending_events();
+	assert_eq!(events.len(), 1);
+	match events[0] {
+		Event::InvoiceRequestFailed { payment_id: pay_id } => {
+			assert_eq!(pay_id, payment_id)
+		},
+		_ => panic!("Unexpected Event"),
+	}
```
Let's use `get_event!` as seen elsewhere in this file.

shaavan replied:
Done! [58f8f6a](https://github.com/lightningdevkit/rust-lightning/commit/58f8f6a8f5d2ac1ea07d45feab933784fa98bb09)
Thanks for the suggestion!

shaavan commented:
Updated from [pr3085.06](https://github.com/shaavan/rust-lightning/commits/pr3085.06) to [pr3085.07](https://github.com/shaavan/rust-lightning/commits/pr3085.07) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3085.06..pr3085.07)):
Addressed @jkczyz comments

Changes:
1. Use `get_event!()` in test to maintain consistency.
2. Reverted formatting in `Readable for ControlTlvs`, to maintain Git Blame.

shaavan commented on lightning/src/onion_message/packet.rs:
```diff
@@ -323,9 +323,7 @@ impl Readable for ControlTlvs {
-			ControlTlvs::Receive(ReceiveTlvs {
-				path_id,
-			})
+			ControlTlvs::Receive(ReceiveTlvs { path_id })
```
Done! [58f8f6a](https://github.com/lightningdevkit/rust-lightning/commit/58f8f6a8f5d2ac1ea07d45feab933784fa98bb09)
Thanks for the pointer!

shaavan commented:
Updated from [pr3085.07](https://github.com/shaavan/rust-lightning/commits/pr3085.07) to [pr3085.08](https://github.com/shaavan/rust-lightning/commits/pr3085.08) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3085.07..pr3085.08)):

Changes:
1. Rebase on main to resolve merge conflicts.

jkczyz commented on lightning/src/blinded_path/message.rs:
```diff
@@ -13,6 +13,8 @@
 
 use bitcoin::secp256k1::{self, PublicKey, Secp256k1, SecretKey};
 
+use crate::ln::channelmanager::PaymentId;
+use crate::ln::msgs::DecodeError;
```
Could you move these into the import block below, before `crate::ln::onion_utils`?

shaavan replied:
Moved the imports to the appropriate location in [6ff2e49](https://github.com/lightningdevkit/rust-lightning/commit/6ff2e49123c3135dccea720d7104e6ad5b8f7308)
Thanks!

jkczyz commented on lightning/src/blinded_path/message.rs:
```diff
@@ -52,12 +54,27 @@ pub(crate) struct ForwardTlvs {
 
 /// Similar to [`ForwardTlvs`], but these TLVs are for the final node.
 pub(crate) struct ReceiveTlvs {
-	/// If `path_id` is `Some`, it is used to identify the blinded path that this onion message is
-	/// sending to. This is useful for receivers to check that said blinded path is being used in
-	/// the right context.
```
Let's keep this comment.

shaavan replied:
Done! [6ff2e49](https://github.com/lightningdevkit/rust-lightning/commit/6ff2e49123c3135dccea720d7104e6ad5b8f7308)
Thanks!

jkczyz commented on lightning/src/blinded_path/message.rs:
```diff
@@ -74,16 +91,39 @@ impl Writeable for ForwardTlvs {
 
-impl Writeable for ReceiveTlvs {
-	fn write<W: Writer>(&self, writer: &mut W) -> Result<(), io::Error> {
-		// TODO: write padding
```
Let's keep this comment. Also, could you move the new code back to here to keep the blame layer clean?

shaavan replied:
Done! [6ff2e49](https://github.com/lightningdevkit/rust-lightning/commit/6ff2e49123c3135dccea720d7104e6ad5b8f7308)
Thanks for the pointer!

jkczyz commented on lightning/src/blinded_path/message.rs:
```diff
@@ -74,16 +91,39 @@ impl Writeable for ForwardTlvs {
-		});
-		Ok(())
-	}
+/// Represents additional data included by the recipient in a [BlindedPath].
+///
+/// This data is encrypted and will be included when sending through
+/// [BlindedPath]. The recipient can authenticate the message and
```
Use ticks around identifiers. Seems github converted into formatting when I suggested the change.

shaavan replied:
Done! [653636b](https://github.com/lightningdevkit/rust-lightning/commit/653636b521ec0d4dba73b7aff83e80d594928a9e)
Thanks for catching it!

jkczyz commented on lightning/src/blinded_path/message.rs:
```diff
@@ -74,16 +91,39 @@ impl Writeable for ForwardTlvs {
+///
+/// [`OffersMessage`]: crate::onion_message::offers::OffersMessage
+#[derive(Clone, Debug)]
+pub struct OffersData {
+	/// Payment ID of the outbound BOLT12 payment.
+	pub payment_id: Option<PaymentId>
 }
```
We're going to need different variants for storing data in offer paths (see #3117), so let's go ahead and make this an enum and serialize using the corresponding macros. Probably can make an empty variant at least temporarily and another variant with a non-`Option` payment id.

shaavan replied:
Makes sense!
Converted the OffersData into an enum in [653636b](https://github.com/lightningdevkit/rust-lightning/commit/653636b521ec0d4dba73b7aff83e80d594928a9e)
Thanks for the suggestion!

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -931,9 +931,20 @@ where
 	) {
 		Ok((Payload::Receive::<ParsedOnionMessageContents<<<CMH as Deref>::Target as CustomOnionMessageHandler>::CustomMessage>> {
-			message, control_tlvs: ReceiveControlTlvs::Unblinded(ReceiveTlvs { path_id }), reply_path,
+			message, control_tlvs: ReceiveControlTlvs::Unblinded(ReceiveTlvs {path_id}), reply_path,
```
Keep the spaces

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -1413,11 +1424,11 @@ where
 		let logger = WithContext::from(&self.logger, Some(*peer_node_id), None, None);
 		match self.peel_onion_message(msg) {
-			Ok(PeeledOnion::Receive(message, path_id, reply_path)) => {
+			Ok(PeeledOnion::Receive(message, _tlvs, reply_path)) => {
```
Let's call this `_recipient_data`

shaavan replied:
Alright! [653636b](https://github.com/lightningdevkit/rust-lightning/commit/653636b521ec0d4dba73b7aff83e80d594928a9e)

jkczyz commented on lightning/src/blinded_path/message.rs:
```diff
@@ -139,7 +139,7 @@ pub(super) fn blinded_hops<T: secp256k1::Signing + secp256k1::Verification>(
 		})
 		.map(|next_hop| ControlTlvs::Forward(ForwardTlvs { next_hop, next_blinding_override: None }))
-		.chain(core::iter::once(ControlTlvs::Receive(ReceiveTlvs { path_id: None })));
+		.chain(core::iter::once(ControlTlvs::Receive(ReceiveTlvs{path_id: recipient_data})));
```
Keep spaces

shaavan replied:
Done! [653636b](https://github.com/lightningdevkit/rust-lightning/commit/653636b521ec0d4dba73b7aff83e80d594928a9e)
Thanks!

shaavan commented:
Updated from [pr3085.08](https://github.com/shaavan/rust-lightning/commits/pr3085.08) to [pr3085.09](https://github.com/shaavan/rust-lightning/commits/pr3085.09) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3085.08..pr3085.09)):
Addressed @jkczyz's comments.

Changes:
1. Reordered imports to match lexicographic ordering.
2. Improve comments, and documentation, to be more clear, correct, and descriptive.
3. Clean the code, and reorder function definition to preserve Git Blame.
4. Rename `tlvs` -> `recipient_data`.
5. Convert `OffersData` in an enum.

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -933,7 +933,18 @@ where
+				(ParsedOnionMessageContents::Custom(_), Some(RecipientData::CustomContext(_))) => {
+					Ok(PeeledOnion::Receive(message, path_id, reply_path))
+				}
+				_ => Err(())
```
Let's also `log_trace` indicating the message was sent on a blinded path with the wrong context. Ideally, indicating the message and context types. Not sure if we want to include the `Debug` content given this would typically be a result of malicious behavior.

shaavan replied:
Introduce a `log_trace!()` for the wrong context in [e4d4218](https://github.com/lightningdevkit/rust-lightning/commit/e4d4218954dbe9564ed1f2ded56f7af515e2950f).

Including `Debug` content could be relevant if the issue arose from us creating a blinded path with the wrong context when we initially provided our peer with a blinded path to respond with. However, since this issue can also result from malicious or faulty peer behavior, assuming our fault through a `debug_assert!()` might not be the best choice in this case.

jkczyz commented on lightning/src/blinded_path/message.rs:
```diff
@@ -84,10 +85,54 @@ impl Writeable for ReceiveTlvs {
+pub enum OffersData {
+	/// Payment ID of the outbound BOLT12 payment.
+	WithPaymentId(PaymentId),
+	Empty(()),
```
Looks like CI is failing because this needs docs.

shaavan replied:
Introduced documentation and other fuzz test fixes in [e4d4218](https://github.com/lightningdevkit/rust-lightning/commit/e4d4218954dbe9564ed1f2ded56f7af515e2950f)
Thanks for the pointer!

shaavan commented:
Updated from [pr3085.09](https://github.com/shaavan/rust-lightning/commits/pr3085.09) to [pr3085.10](https://github.com/shaavan/rust-lightning/commits/pr3085.10) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3085.09..pr3085.10)):

Addressed @jkczyz's comments:

1. Renamed `RecipientData` to MessageContext and OffersData to OffersContext, along with all subsequent variable name changes.
2. Renamed the `OffersContext` variant and improved its ordering for clearer naming and usage.
3. Introduced appropriate documentation and made minor changes to fuzz tests to fix CI issues.
4. Added a `log_trace!` when receiving a message on a `reply_path` with mismatched context.

shaavan commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -1430,14 +1422,14 @@ where
 						let responder = reply_path.map(
-							|reply_path| Responder::new(reply_path, path_id)
+							|reply_path| Responder::new(reply_path)
 						);
```
Wow! Rust is magic! 🪄
Implemented your suggestion in [e4d4218](https://github.com/lightningdevkit/rust-lightning/commit/e4d4218954dbe9564ed1f2ded56f7af515e2950f)

shaavan commented on lightning/src/blinded_path/message.rs:
```diff
@@ -12,19 +12,20 @@
 //! [`BlindedPath`]: crate::blinded_path::BlindedPath
 
 use bitcoin::secp256k1::{self, PublicKey, Secp256k1, SecretKey};
-
```
Done! [e4d4218](https://github.com/lightningdevkit/rust-lightning/commit/e4d4218954dbe9564ed1f2ded56f7af515e2950f)

shaavan commented on lightning/src/blinded_path/message.rs:
```diff
@@ -84,6 +85,50 @@ impl Writeable for ReceiveTlvs {
+/// [`BlindedPath`]. The recipient can authenticate the message and
+/// utilize it for further processing if needed.
+#[derive(Clone, Debug)]
+pub enum RecipientData {
```
Thanks for the naming suggestions!
These sets of names sound much better than the original.
Introduced the new names in [e4d4218](https://github.com/lightningdevkit/rust-lightning/commit/e4d4218954dbe9564ed1f2ded56f7af515e2950f)
Thanks!

shaavan commented on lightning/src/blinded_path/message.rs:
```diff
@@ -84,6 +85,50 @@ impl Writeable for ReceiveTlvs {
+pub enum OffersData {
+	/// Payment ID of the outbound BOLT12 payment.
+	WithPaymentId(PaymentId),
+	Empty(()),
```
Done!
Renamed and Reorder the Variants and introduce struct syntax in [e4d4218](https://github.com/lightningdevkit/rust-lightning/commit/e4d4218954dbe9564ed1f2ded56f7af515e2950f)
Thanks for the suggestion!

shaavan commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -933,7 +933,18 @@ where
 			message, control_tlvs: ReceiveControlTlvs::Unblinded(ReceiveTlvs { path_id }), reply_path,
 		}, None)) => {
-			Ok(PeeledOnion::Receive(message, path_id, reply_path))
+			match (&message, &path_id) {
```
Got it! Only `ReceiveTlvs` will know about `path_id` :P
[e4d4218](https://github.com/lightningdevkit/rust-lightning/commit/e4d4218954dbe9564ed1f2ded56f7af515e2950f)

shaavan commented on fuzz/src/onion_message.rs:
```diff
@@ -99,7 +100,7 @@ struct TestOffersMessageHandler {}
 impl OffersMessageHandler for TestOffersMessageHandler {
 	fn handle_message(
-		&self, _message: OffersMessage, _responder: Option<Responder>,
+		&self, _message: OffersMessage, _responder: Option<Responder>, _offers_data: OffersData,
```
Done! [e4d4218](https://github.com/lightningdevkit/rust-lightning/commit/e4d4218954dbe9564ed1f2ded56f7af515e2950f)
Thanks!

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -1436,12 +1436,30 @@ where
+						let context = match context {
+							None => OffersContext::Unknown {},
+							Some(MessageContext::Offers(context)) => context,
+							Some(MessageContext::Custom(_)) => {
+								debug_assert!(false, "Shouldn't have triggered this case.");
+								return
+							}
```
Could you remind me why we can't do this as part of `Payload` parsing? I think we discussed it, but it would be nice to avoid this case if we can handle it earlier.

shaavan replied:
So the way we have set it up ensures that only the correct (`msg_type`, `recipient_data_type`) combinations are parsed in `peel_onion_message`. The checks we're performing in `handle_onion_message` serve as an additional fail-safe to ensure that this scenario isn’t ever triggered, whether due to a mistake on our end or a malicious attempt by someone else.

One quick question: since we expect this condition to never actually occur, do you think we should implement a stricter response, such as using an explicit `panic!()`?
Let me know what you think!

Thanks!

jkczyz replied:
> So the way we have set it up ensures that only the correct (`msg_type`, `recipient_data_type`) combinations are parsed in `peel_onion_message`. The checks we're performing in `handle_onion_message` serve as an additional fail-safe to ensure that this scenario isn’t ever triggered, whether due to a mistake on our end or a malicious attempt by someone else.

Right, but that doesn't answer my question. More precisely, was there a reason why we aren't able to include the context data as another part of `ParsedOnionMessageContents`? (i.e. have each variant wrap both a message and a context instead of having the context returned separately). That would avoid the problem entirely.

Here we are returning a `Payload<ParsedOnionMessageContents<..>>` which spreads the the message and context across two different fields, one being an enum with a variant containing the `ReceiveTlvs`. Maybe that is why?

https://github.com/lightningdevkit/rust-lightning/blob/88e1b56d66ff550b36a6d422f47c9b9729406f61/lightning/src/onion_message/packet.rs#L243

@TheBlueMatt Any opinon on this? Would it be worth refactoring (if not now, then in a follow-up)?

jkczyz replied:
Spoke offline. Let's leave this to a follow-up.

shaavan commented:
Updated from [pr3085.10](https://github.com/shaavan/rust-lightning/commits/pr3085.10) to [pr3085.11](https://github.com/shaavan/rust-lightning/commits/pr3085.11) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3085.10..pr3085.11)):
Addressed @jkczyz comment

1. Expanded the match statement in handle_message, to make the `Err`, and `Ok` cases explicit.
2. Expanded the log message to also include the error itself.

shaavan commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -10314,18 +10322,24 @@ where
+						abandon_if_payment(context);
+						log_trace!(
+							self.logger,
+							"A response was generated, but there is no reply_path specified for sending the response."
```
Sounds like a good idea!
Addressed in [c81916e](https://github.com/lightningdevkit/rust-lightning/commit/c81916efb896e0011749bcd7039c75ef1ac26efd)
Thanks for the suggestion!

shaavan commented:
Updated from [pr3085.11](https://github.com/shaavan/rust-lightning/commits/pr3085.11) to [pr3085.12](https://github.com/shaavan/rust-lightning/commits/pr3085.12) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3085.11..pr3085.12)):

Changes:
1. Rebase on main, to resolve merge conflicts.

jkczyz commented on lightning/src/blinded_path/message.rs:
```diff
@@ -84,6 +85,57 @@ impl Writeable for ReceiveTlvs {
+		/// Payment ID of the outbound BOLT12 payment.
+		payment_id: PaymentId
+	},
+
```
nit: remove new line

shaavan replied:
Done! [3285830](https://github.com/lightningdevkit/rust-lightning/commit/328583045b9aadb1e35e6d3b180d0a30ca526a63)
Thanks!

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -10367,14 +10375,25 @@ where
-							ResponseInstruction::NoResponse
-						},
-					},
+					Err(err) => {
+						match responder {
+							Some(responder) => {
+								abandon_if_payment(context);
+								responder.respond(OffersMessage::InvoiceError(err))
+							}
+							None => {
+								abandon_if_payment(context);
+								log_trace!(
+									self.logger,
+									"A error response was generated, but there is no reply_path specified \
+									for sending the response. Error: {}",
+									err
+								);
+								return ResponseInstruction::NoResponse;
+							}
+						}
+					}
```
Since this arm uses a single expression -- the `match` -- it can be moved up in order to reduce the level indentation.

```rust
Err(err) => match responder {
	// ...
},

shaavan replied:
Done! Cleaned the code in [3285830](https://github.com/lightningdevkit/rust-lightning/commit/328583045b9aadb1e35e6d3b180d0a30ca526a63)
Thanks for the suggestion!

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -10367,14 +10375,25 @@ where
+									for sending the response. Error: {}",
+									err
+								);
+								return ResponseInstruction::NoResponse;
+							}
+						}
+					}
+					Ok(_) => return ResponseInstruction::NoResponse,
```
No need for explicit returns here when the expression evaluates to the return value.

jkczyz commented on lightning/src/blinded_path/message.rs:
```diff
@@ -84,10 +86,61 @@ impl Writeable for ReceiveTlvs {
+impl Readable for ReceiveTlvs {
+	fn read<R: io::Read>(r: &mut R) -> Result<Self, DecodeError> {
+		_init_and_read_tlv_stream!(r, {
+			(6, path_id, option),
```
Spoke with @TheBlueMatt offline. Let's use a higher number like we do for payment `ReceiveTlvs` instead of using `path_id`. That way if we ever want to create blinded paths for other implementations, we can do so using `path_id`. Also, rename `path_id` to `context` in `ReceiveTlvs`.

shaavan commented:
Updated from [pr3085.12](https://github.com/shaavan/rust-lightning/commits/pr3085.12) to [pr3085.13](https://github.com/shaavan/rust-lightning/commits/pr3085.13) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3085.12..pr3085.13)):
Addressed @jkczyz comments

Changes:
1. Rename `path_id` -> `context` in ReceiveTlvs, and use `65537` as the tlv number.
2. Remove extra line and indentations.

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -795,8 +795,8 @@ pub trait CustomOnionMessageHandler {
 	Forward(NextMessageHop, OnionMessage),
-	/// Received onion message, with decrypted contents, path_id, and reply path
-	Receive(ParsedOnionMessageContents<T>, Option<[u8; 32]>, Option<BlindedPath>)
+	/// Received onion message, with decrypted contents, recipient data, and reply path
```
s/recipient data/context

shaavan replied:
Done! [528dc64](https://github.com/lightningdevkit/rust-lightning/commit/528dc644e5b995c10f96fe36fb803bc6433580d7)
Thanks for the pointer!

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -1432,11 +1446,11 @@ where
 		let logger = WithContext::from(&self.logger, Some(*peer_node_id), None, None);
 		match self.peel_onion_message(msg) {
-			Ok(PeeledOnion::Receive(message, path_id, reply_path)) => {
+			Ok(PeeledOnion::Receive(message, _recipient_data, reply_path)) => {
```
s/_recipient_data/_context

shaavan replied:
Done! [528dc64](https://github.com/lightningdevkit/rust-lightning/commit/528dc644e5b995c10f96fe36fb803bc6433580d7)
Thanks!

jkczyz commented on lightning/src/onion_message/packet.rs:
```diff
@@ -331,7 +331,7 @@ impl Readable for ControlTlvs {
 		};
 
-		let valid_fwd_fmt = next_hop.is_some() && path_id.is_none();
+		let valid_fwd_fmt = next_hop.is_some() && context.is_none();
```
@TheBlueMatt The second check is only relevant for `path_id`, which since we now don't understand would result in us failing to parse (i.e., ignoring) this message as dictated by the spec:

https://github.com/lightning/bolts/blob/c562d91ace0e95bec3c6f8758969eaf3627f23c8/04-onion-routing.md?plain=1#L1532-L1536

So we should probably drop the `context` check?

shaavan replied:
I have kept the check intact for now in [528dc64](https://github.com/lightningdevkit/rust-lightning/commit/528dc644e5b995c10f96fe36fb803bc6433580d7).
Waiting eagerly for @TheBlueMatt's thoughts on this before I remove it!

shaavan replied:
After our offline discussion I have removed the check in [9d7a3bf](https://github.com/lightningdevkit/rust-lightning/commit/9d7a3bf8f86fbb64354ffe5511719900a134c668)
Thanks for the suggestion and discussion!

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -10251,10 +10258,17 @@ where
 
+		let abandon_if_payment = |context| {
+			match context {
+				OffersContext::OutboundPayment { payment_id } => self.abandon_payment(payment_id),
```
Yeah, I'll do it as part of #3139. We'll just check that the `payment_id` decrypted from the verified invoice matches this. Though we can't really do the same when handling an `InvoiceError`.

jkczyz replied:
We may need to engage on the spec then, since `InvoiceError` doesn't contain TLVs from the preceding messages.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -10367,14 +10375,23 @@ where
+							abandon_if_payment(context);
+							log_trace!(
+								self.logger,
+								"A error response was generated, but there is no reply_path specified \
```
s/A/An

shaavan replied:
Done! [528dc64](https://github.com/lightningdevkit/rust-lightning/commit/528dc644e5b995c10f96fe36fb803bc6433580d7)
Must have missed it while I fixed the other log message 😅
Thanks for pointing it out!

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -10367,14 +10375,23 @@ where
+							return ResponseInstruction::NoResponse;
+						}
+					}
+					Ok(_) => ResponseInstruction::NoResponse,
```
Let's put the `Ok` arm before the `Err` arm.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -10367,14 +10375,23 @@ where
+						Some(responder) => {
+							abandon_if_payment(context);
+							responder.respond(OffersMessage::InvoiceError(err))
+						}
```
nit: here and below, end match arms with a comma.

shaavan replied:
Added proper commas in [528dc64](https://github.com/lightningdevkit/rust-lightning/commit/528dc644e5b995c10f96fe36fb803bc6433580d7)
Thanks for the pointer!

jkczyz commented on lightning/src/onion_message/offers.rs:
```diff
@@ -44,7 +45,7 @@ pub trait OffersMessageHandler {
 	///
 	/// [`OnionMessenger`]: crate::onion_message::messenger::OnionMessenger
-	fn handle_message(&self, message: OffersMessage, responder: Option<Responder>) -> ResponseInstruction<OffersMessage>;
+	fn handle_message(&self, message: OffersMessage, responder: Option<Responder>, context: OffersContext) -> ResponseInstruction<OffersMessage>;
```
Let's move the `context` parameter before the `responder` parameter. Same for the custom handler.

shaavan replied:
Done! Fix the ordering of parameters for `handle_message` and `handle_custom_message` in [528dc64](https://github.com/lightningdevkit/rust-lightning/commit/528dc644e5b995c10f96fe36fb803bc6433580d7)
Thanks!

shaavan commented:
Updated from [pr3085.13](https://github.com/shaavan/rust-lightning/commits/pr3085.13) to [pr3085.14](https://github.com/shaavan/rust-lightning/commits/pr3085.14) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3085.13..pr3085.14)):
Addressed @jkczyz comments

Changes:
1. Reordered the parameters to match the logical sequence in handle_message and handle_custom_message.
2. Fix formatting in `handle_message`.
3. Fix a grammatical mistake in a log message.
4. Fix a comment by renaming `recipient data` -> `context`.

shaavan commented:
Updated from [pr3085.13](https://github.com/shaavan/rust-lightning/commits/pr3085.13) to [pr3085.14](https://github.com/shaavan/rust-lightning/commits/pr3085.14) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3085.13..pr3085.14)):
Addressed @jkczyz comments

Changes:
1. Reordered the parameters to match the logical sequence in handle_message and handle_custom_message.
2. Fix formatting in `handle_message`.
3. Fix a grammatical mistake in a log message.
4. Fix a comment by renaming `recipient data` -> `context`.

shaavan commented:
Updated from [pr3085.14](https://github.com/shaavan/rust-lightning/commits/pr3085.14) to [pr3085.15](https://github.com/shaavan/rust-lightning/commits/pr3085.15) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3085.14..pr3085.15)):
Addressed @jkczyz comments

1. Remove `context.is_none()` check from ControlTlvs' Readable.
2. Replace `recipient_data` with `context` from renaming functions, and fix commit messages.
3. Update `create_blinded_path` functions to take in MessageContext instead of Option<MessageContext>.
4. Update ResponseInstruction::WithReplyPath to contain and send context with response.

shaavan commented on fuzz/src/chanmon_consistency.rs:
```diff
@@ -138,7 +139,8 @@ impl MessageRouter for FuzzRouter {
 
 	fn create_blinded_paths<T: secp256k1::Signing + secp256k1::Verification>(
-		&self, _recipient: PublicKey, _peers: Vec<PublicKey>, _secp_ctx: &Secp256k1<T>,
+		&self, _recipient: PublicKey, _recipient_data: Option<MessageContext>,
```
Done! [9d7a3bf](https://github.com/lightningdevkit/rust-lightning/commit/9d7a3bf8f86fbb64354ffe5511719900a134c668)
Thanks for the pointer!

shaavan commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -1178,7 +1183,7 @@ where
 
 		self.message_router
-			.create_blinded_paths(recipient, peers, secp_ctx)
+			.create_blinded_paths(recipient, None, peers, secp_ctx)
```
Updated the ResponseInstruction::WithReplyPath() and `messenger.rs`' create_blinded_path to take MessageContext as input field in [9d7a3bf](https://github.com/lightningdevkit/rust-lightning/commit/9d7a3bf8f86fbb64354ffe5511719900a134c668)
This will allow us to pass the context here!
Thanks for the suggestion!

shaavan commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -8375,7 +8375,7 @@ macro_rules! create_offer_builder { ($self: ident, $builder: ty) => {
 		let secp_ctx = &$self.secp_ctx;
 
-		let path = $self.create_blinded_path_using_absolute_expiry(absolute_expiry)
+		let path = $self.create_blinded_path_using_absolute_expiry(absolute_expiry, None)
```
Makes sense!
Updated in [9d7a3bf](https://github.com/lightningdevkit/rust-lightning/commit/9d7a3bf8f86fbb64354ffe5511719900a134c668)
Thanks!

shaavan commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -441,7 +442,7 @@ pub trait MessageRouter {
 		T: secp256k1::Signing + secp256k1::Verification
 	>(
-		&self, recipient: PublicKey, peers: Vec<PublicKey>, secp_ctx: &Secp256k1<T>,
+		&self, recipient: PublicKey, context: Option<MessageContext>, peers: Vec<PublicKey>, secp_ctx: &Secp256k1<T>,
```
I don't think there will be a case where we have to pass None.
Updated create_blinded_paths to take MessageContext as input in [9d7a3bf](https://github.com/lightningdevkit/rust-lightning/commit/9d7a3bf8f86fbb64354ffe5511719900a134c668)

shaavan commented:
Updated from [pr3085.14](https://github.com/shaavan/rust-lightning/commits/pr3085.14) to [pr3085.15](https://github.com/shaavan/rust-lightning/commits/pr3085.15) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3085.14..pr3085.15)):
Addressed @jkczyz comments

1. Remove `context.is_none()` check from ControlTlvs' Readable.
2. Replace `recipient_data` with `context` from renaming functions, and fix commit messages.
3. Update `create_blinded_path` functions to take in MessageContext instead of Option<MessageContext>.
4. Update ResponseInstruction::WithReplyPath to contain and send context with the response.

shaavan commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -8803,15 +8809,15 @@ where
 	/// [`MAX_SHORT_LIVED_RELATIVE_EXPIRY`].
 	fn create_blinded_path_using_absolute_expiry(
-		&self, absolute_expiry: Option<Duration>
+		&self, absolute_expiry: Option<Duration>, context: Option<MessageContext>
```
Addressed in [f243a9a](https://github.com/lightningdevkit/rust-lightning/commit/f243a9a6974fd2558ccd6853dd083a8346560e06)
Thanks for the suggestion!

shaavan commented:
Updated from [pr3085.15](https://github.com/shaavan/rust-lightning/commits/pr3085.15) to [pr3085.16](https://github.com/shaavan/rust-lightning/commits/pr3085.16) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3085.15..pr3085.16)):

Changes:
1. Fix fuzz test.

jkczyz commented on lightning/src/blinded_path/message.rs:
```diff
@@ -78,12 +79,62 @@ impl Writeable for ReceiveTlvs {
 	}
 }
 
+impl Readable for ReceiveTlvs {
+	fn read<R: io::Read>(r: &mut R) -> Result<Self, DecodeError> {
+		_init_and_read_tlv_stream!(r, {
+			(65537, context, option),
+		});
+		Ok(ReceiveTlvs { context })
+	}
+}
+
```
Is this needed? We are reading `ReceiveTlvs` indirectly when reading `ControlTlvs`.

shaavan replied:
Just check that it is redundant!
Removed in [c8de6b6](https://github.com/lightningdevkit/rust-lightning/commit/c8de6b610fdde71730c5f536ecd89e9afcf003ea)
Thanks for the suggestion!

jkczyz commented on fuzz/src/invoice_request_deser.rs:
```diff
@@ -87,10 +87,22 @@ fn build_response<T: secp256k1::Signing + secp256k1::Verification>(
-			.unwrap(),
-		BlindedPath::new_for_message(&intermediate_nodes[1], pubkey(42), &entropy_source, secp_ctx)
-			.unwrap(),
+		BlindedPath::new_for_message(
+			&intermediate_nodes[0],
+			pubkey(42),
+			MessageContext::Custom(Vec::new()),
+			&entropy_source,
+			secp_ctx,
+		)
+		.unwrap(),
+		BlindedPath::new_for_message(
+			&intermediate_nodes[1],
+			pubkey(42),
+			MessageContext::Custom(Vec::new()),
+			&entropy_source,
+			secp_ctx,
+		)
```
Let's use `OffersContext::Unknown` here.

jkczyz commented on fuzz/src/refund_deser.rs:
```diff
@@ -76,10 +76,22 @@ fn build_response<T: secp256k1::Signing + secp256k1::Verification>(
-			.unwrap(),
-		BlindedPath::new_for_message(&intermediate_nodes[1], pubkey(42), &entropy_source, secp_ctx)
-			.unwrap(),
+		BlindedPath::new_for_message(
+			&intermediate_nodes[0],
+			pubkey(42),
+			MessageContext::Custom(Vec::new()),
+			&entropy_source,
+			secp_ctx,
+		)
+		.unwrap(),
+		BlindedPath::new_for_message(
+			&intermediate_nodes[1],
+			pubkey(42),
+			MessageContext::Custom(Vec::new()),
+			&entropy_source,
+			secp_ctx,
+		)
+		.unwrap(),
```
Same here.

shaavan replied:
Done! [c8de6b6](https://github.com/lightningdevkit/rust-lightning/commit/c8de6b610fdde71730c5f536ecd89e9afcf003ea)
Thanks!

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -1267,15 +1272,15 @@ where
 		let message_type = response.message.msg_type();
 		let reply_path = if create_reply_path {
-			match self.create_blinded_path() {
+			match self.create_blinded_path(context.unwrap()) {
```
Instead of keeping `create_reply_path`, let's drop it in favor of using `context`. That way we don't need to `unwrap` here, which is best to avoid.

shaavan replied:
Makes sense!
Updated in [c8de6b6](https://github.com/lightningdevkit/rust-lightning/commit/c8de6b610fdde71730c5f536ecd89e9afcf003ea)
Thanks for the pointer!

jkczyz reviewed (commented):
Pretty much LGTM other than a few last comments.

shaavan commented:
Updated from [pr3085.16](https://github.com/shaavan/rust-lightning/commits/pr3085.16) to [pr3085.17](https://github.com/shaavan/rust-lightning/commits/pr3085.17) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3085.16..pr3085.17)):
Addressed @jkczyz, @TheBlueMatt comments

1. Update the fuzz test to take in the right MessageContext.
2. Remove redundant Readable for ReceiveTlvs implementation.
3. Update `handle_onion_message_response` to check based on `context` instead of boolean.
4. Update documentation.

shaavan commented on lightning/src/blinded_path/message.rs:
```diff
@@ -55,7 +57,7 @@ pub(crate) struct ReceiveTlvs {
 	/// sending to. This is useful for receivers to check that said blinded path is being used in
 	/// the right context.
-	pub(crate) path_id: Option<[u8; 32]>,
+	pub context: Option<MessageContext>
```
Done! [c8de6b6](https://github.com/lightningdevkit/rust-lightning/commit/c8de6b610fdde71730c5f536ecd89e9afcf003ea)
Thanks for the pointer!

shaavan commented on lightning/src/blinded_path/message.rs:
```diff
@@ -78,16 +80,66 @@ impl Writeable for ReceiveTlvs {
+/// [`OffersMessage`]: crate::onion_message::offers::OffersMessage
+#[derive(Clone, Debug)]
+pub enum OffersContext {
+	/// Represents an unknown context.
```
Updated the documentation to be more clear in [c8de6b6](https://github.com/lightningdevkit/rust-lightning/commit/c8de6b610fdde71730c5f536ecd89e9afcf003ea)
Let me know if it looks good!
Thanks so much for the pointer!

shaavan commented on lightning/src/blinded_path/message.rs:
```diff
@@ -78,16 +80,66 @@ impl Writeable for ReceiveTlvs {
+/// Represents additional data included by the recipient in a [`BlindedPath`].
+///
+/// This data is encrypted and will be included when sending through
+/// [`BlindedPath`]. The recipient can authenticate the message and
```
Thanks for the suggestion!
Updated the docs in [c8de6b6](https://github.com/lightningdevkit/rust-lightning/commit/c8de6b610fdde71730c5f536ecd89e9afcf003ea)!

shaavan commented on lightning/src/blinded_path/message.rs:
```diff
@@ -78,16 +80,66 @@ impl Writeable for ReceiveTlvs {
+	///
+	/// [`OffersMessage`]: crate::onion_message::offers::OffersMessage
+	Offers(OffersContext),
+	/// Represents the data appended with Custom Onion Message.
```
Done! [c8de6b6](https://github.com/lightningdevkit/rust-lightning/commit/c8de6b610fdde71730c5f536ecd89e9afcf003ea)
Thanks!

jkczyz commented on lightning/src/blinded_path/message.rs:
```diff
@@ -78,16 +79,61 @@ impl Writeable for ReceiveTlvs {
+/// Represents additional data included by the recipient in a [`BlindedPath`].
+///
+/// This data is encrypted by the recipient and remains invisible to anyone else.
+/// It is included in the [BlindedPath], making it accessible again to the recipient
+/// whenever the [BlindedPath] is used.
```
Use ticks around `BlindedPath` so it is formatted.

shaavan replied:
Done! [8137b53](https://github.com/lightningdevkit/rust-lightning/commit/8137b53b96d4e7104bde8533b50d622230751ba2)
Thanks!

jkczyz commented on lightning/src/blinded_path/message.rs:
```diff
@@ -78,16 +79,61 @@ impl Writeable for ReceiveTlvs {
+#[derive(Clone, Debug)]
+pub enum OffersContext {
+	/// Represents an unknown BOLT12 payment context.
+	/// This variant is used when we are unsure if we are dealing
+	/// with an Inbound or Outbound BOLT12 payment.
```
Hmm... how about:

```rust
/// This variant is used when a message is sent without using a [`BlindedPath`] or over one created prior to LDK version 0.0.124.
```

shaavan replied:
Well, this one's much better! 😄
Updated in [8137b53](https://github.com/lightningdevkit/rust-lightning/commit/8137b53b96d4e7104bde8533b50d622230751ba2)
Thanks for the suggestion!

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -1466,7 +1483,15 @@ where
 					ParsedOnionMessageContents::Custom(msg) => {
-						let response_instructions = self.custom_handler.handle_custom_message(msg, responder);
+						let context = match context {
+							None => vec![],
```
Should we even accept a custom message that doesn't have a context? I guess that would assume blinded paths MUST be used, though.

jkczyz replied:
Yeah, makes sense. @shaavan Let's make this an `Option`.

shaavan commented on lightning/src/blinded_path/message.rs:
```diff
@@ -78,16 +80,66 @@ impl Writeable for ReceiveTlvs {
+	},
+}
+
+impl_writeable_tlv_based_enum!(MessageContext, ;
```
Thanks for pointing this out! I think I understand the issue you're describing.

That said, `MessageContext` variants are designed to hold only one struct, which corresponds to their specific `Context` types. Switching to a named-field version might make the code a bit more verbose.

I agree that implementing a `_v2` could be a more robust, long-term fix for this bug. It seems like a cleaner solution overall.

I'd love to hear your thoughts on this. If you still prefer to go with the named-field version, I'm happy to make that change. Let me know what you think!

Thanks a lot, Matt!

shaavan commented:
Updated from [pr3085.17](https://github.com/shaavan/rust-lightning/commits/pr3085.17) to [pr3085.18](https://github.com/shaavan/rust-lightning/commits/pr3085.18) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3085.17..pr3085.18)):
Addressed @jkczyz comments

Update:
1. Fix documentation
2. Fix ci.

jkczyz commented:
CI is sad because `ln::max_payment_path_len_tests::bolt12_invoice_too_large_blinded_paths` is now failing. @shaavan Would you mind rebasing and updating that test accordingly now that an `PaymentFailed` event is generated?

shaavan commented:
Updated from [pr3085.18](https://github.com/shaavan/rust-lightning/commits/pr3085.18) to [pr3085.19](https://github.com/shaavan/rust-lightning/commits/pr3085.19) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3085.18..pr3085.19)):
Addressed @jkczyz comments

1. Rebase on main, and fix the `bolt12_invoice_too_large_blinded_paths` test, to make CI happy :)

Thanks for the pointer, @jkczyz!

shaavan commented:
Updated from [pr3085.19](https://github.com/shaavan/rust-lightning/commits/pr3085.19) to [pr3085.20](https://github.com/shaavan/rust-lightning/commits/pr3085.20) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3085.19..pr3085.20)):
Addressed @TheBlueMatt and @jkczyz comment

1. Update `handle_custom_message` input field to take in `Option<Vec>` instead of `<Vec>`.

jkczyz reviewed (approved):
Please update the PR title and summary to reflect name changes.

jkczyz commented on lightning/src/blinded_path/message.rs:
```diff
@@ -78,16 +79,62 @@ impl Writeable for ReceiveTlvs {
+/// This data is encrypted by the recipient and remains invisible to anyone else.
+/// It is included in the [`BlindedPath`], making it accessible again to the recipient
+/// whenever the [`BlindedPath`] is used.
+/// The recipient can authenticate the message and utilize it for further processing
```
I'll update this in #3139 since I'm touching this.
